crates/subspace-service/src/dsn.rs

```diff
@@ -46,13 +47,24 @@ where
             }) = archived_segment_notification_stream.next().await
             {
                 trace!(target: "dsn", "ArchivedSegmentNotification received");
-                let data = archived_segment.encode().to_vec();
 
-                match node.publish(PUB_SUB_ARCHIVING_TOPIC.clone(), data).await {
-                    Ok(_) => {
+                let segment_index = archived_segment.root_block.segment_index();
+                let first_piece_index = segment_index * u64::from(PIECES_IN_SEGMENT);
+
+                let keys_iter = (first_piece_index..)
+                    .take(archived_segment.pieces.count())
+                    .map(PieceIndexHash::from_index)
+                    .map(|hash| hash.into());
+
+                let pieces_announcements = keys_iter
+                    .map(|key| node.announce_piece(key))
+                    .collect::<Vec<_>>();
+                let announcement_result = join_all(pieces_announcements).await;
```

Comment:
You can collect into https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html instead and await on that, should be faster. Or there is an ordered one too if order is important.